File: c++/cote/GJM/사이클_존재여부_확인하기.cpp
#include <iostream>
#include <random>
#include <random>
#define TEST_CASES 10
#define N 1000

using namespace std;

vector<vector<bool>> initialize(default_random_engine& gen) {
	int n = gen() % (N - 3) + 3;
	vector<vector<bool>> ret (n, vector<bool>(n, false));

	for (int i = 0; i < n; i++) {
		for (int j = 0; j < n; j++) {
			if (i == j || gen() % 10000 < 9990 ) {
				continue;
			}
			ret[i][j] = true;
		}
	}

	return ret;
}

// 역방향 간선이 있으면 사이클이 존재하는 지 확인하기 위해, 깊이 탐색 path에 자신의 outgoing 간선이 있는지 확인한다.  
bool dfs(int cur, vector<vector<bool>>& graph, vector<bool>& visited, vector<int> path) {
	visited[cur] = true;
	path.push_back(cur);

	for (int next = 0; next < graph.size(); next++){
		if (graph[cur][next]) {
			if (!visited[next]) {
				if(!dfs(next, graph, visited, path)) {
					return false;
				}
			}	

			if(auto it = find(begin(path), end(path), next); it != end(path)) {
				return false;
			}
		} 
	}
	return true;
}

bool solve(vector<vector<bool>>& graph) {
	vector<bool> visited(graph.size(), false);

	for (int i = 0; i < graph.size(); i++) {
		if (!visited[i]){
			if (!dfs(i, graph, visited, vector<int>{i})) {
				return false;
			}
		}
	}

	return true;
}

int main(void) {	
	default_random_engine gen(43);
	
	for (int it = 0; it < TEST_CASES; it++) {
		vector<vector<bool>> graph = initialize(gen);

		cout << boolalpha << solve(graph) << endl;
	}

	return 0;
}

File: go/cote/GJM/사이클_존재여부_확인하기.go
package main

import (
	"fmt"
	"math/rand"
	"slices"
)

const (
	TEST_CASES = 10
	N = 10
)

func _dfs(cur int, adj [][]int, visited []bool) {
	visited[cur] = true

	for _, next := range adj[cur] {
		if !visited[next] {
			_dfs(next, adj, visited)
		}
	}
}

func check_reachable(adj [][]int) bool {
	visited := make([]bool, len(adj)) 

	_dfs(0, adj, visited)

	for _, v := range visited {
		if !v {
			return false
		} 
	}

	return true
}

func initialize(random *rand.Rand) (ret [][]int) {

	for {
		n := random.Intn(N - 3) + 3
		ret = make([][]int, n)
		for i, _ := range ret {
			for j := 0; j < n; j++ {
				if i == j || random.Float32() < 0.9 {
					continue
				}
				ret[i] = append(ret[i], j)
			}
		}

		if check_reachable(ret)  {
			return
		}

	}
	return
}

func dfs(cur int, adj [][]int, visited []bool, path []int) bool {
	visited[cur] = true

	for _, next := range adj[cur] {
		if !visited[next]  {
			newPath := append(slices.Clone(path), next)
			if dfs(next, adj, visited, newPath) {
				return true
			}

		} else if slices.Contains(path, next){
			return true
		}
	}

	return false
}

func solve(adj [][]int) bool {
	visited := make([]bool, len(adj))

	for i := 0; i < len(adj); i++ {
		if !visited[i] && dfs(i, adj, visited, []int{i}) {
			return true
		}
	}

	return false
}

func main() {
	random := rand.New(rand.NewSource(43))

	for it := 0; it < TEST_CASES; it++ {
		adj := initialize(random)
		fmt.Println(solve(adj))
	}
}